fallback.go

```diff
@@ -0,0 +1,144 @@
+package txtdirect
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+type Fallback struct {
+	rw      http.ResponseWriter
+	request *http.Request
+	config  Config
+
+	records    []record
+	pathRecord record
+	lastRecord record
+
+	fallbackType string
+	code         int
+}
+
+// fallback redirects the request to the given fallback address
+// and if it's not provided it will check txtdirect config for
+// default fallback address
+func fallback(w http.ResponseWriter, r *http.Request, fallbackType string, code int, c Config) {
+	if code == http.StatusMovedPermanently {
+		w.Header().Add("Cache-Control", fmt.Sprintf("max-age=%d", status301CacheAge))
+	}
+	w.Header().Add("Status-Code", strconv.Itoa(code))
+
+	f := Fallback{
+		rw:           w,
+		request:      r,
+		config:       c,
+		fallbackType: fallbackType,
+		code:         code,
+	}
+
+	if fallbackType != "global" {
+		// Fetch records from request's context and set the []record type on them
+		f.fetchRecords()
+
+		// Redirect to first record's `to=` field
+		if (fallbackType == "to" && f.lastRecord.To != "") && f.validateURIs() {
+			http.Redirect(w, r, f.lastRecord.To, code)
+			f.countFallback(f.lastRecord.Type)
+		}
+
+		// Redirect to first record's `website=` field
+		if fallbackType == "website" && f.lastRecord.Website != "" {
+			http.Redirect(w, r, f.lastRecord.Website, code)
+			f.countFallback(f.lastRecord.Type)
+		}
+
+		// Dockerv2 root fallback
+		if fallbackType == "root" && f.lastRecord.Root != "" {
```

Comment:
Why is the root record specific to dockerv2?

Couldn't we in general say the following order makes sense:
to
website
root
website (path)
root (path)
to (path)
global redirect flow

That way the fallback code is not type specific besides having knowledge of path.
The rest would just be per type. So dockerv2 would implement root and website, but gometa would only implement website.

Being on that thought, why does gometa not have root? (just checking my thinking :) )

Reply:
Changed the flow to be like what you've mentioned and removed the `dockerv2` comment.